src/ott/neural/methods/expectile_neural_dual.py

```diff
@@ -0,0 +1,526 @@
+  """The dual potentials of ENOT method.
+
+  Args:
+    grad_f: Gradient of the first dual potential function.
+    g: The second dual potential function.
+    cost_fn: The cost function used to solve the OT problem.
+    is_bidirectional: Whether the duals are trained for bidirectional
+      transport mapping.
+    corr: Whether the duals solve the problem in correlation form.
+  """
+
+  def __init__(
+      self, grad_f: PotentialGradientFn_t, g: PotentialValueFn_t,
+      cost_fn: costs.CostFn, is_bidirectional: bool, corr: bool
+  ):
+
+    self.__grad_f = grad_f
+    self.is_bidirectional = is_bidirectional
+
+    if is_bidirectional and not corr:
+      transport = lambda x: cost_fn.twist_operator(x, grad_f(x), False)
+    else:
+      transport = lambda x: grad_f(x)
+
+    conjugate_cost = jnp.dot if corr else cost_fn
+
+    def g_cost_conjugate(x: jnp.ndarray) -> jnp.ndarray:
+      y_hat = jax.lax.stop_gradient(transport(x))
+      return -g(y_hat) + conjugate_cost(x, y_hat)
+
+    super().__init__(g_cost_conjugate, g, cost_fn=cost_fn, corr=corr)
+
+  @property
+  def _grad_f(self) -> Callable[[jnp.ndarray], jnp.ndarray]:
+    return jax.vmap(self.__grad_f)
+
+  def transport(self, vec: jnp.ndarray, forward: bool = True) -> jnp.ndarray:
+    """Transport from source to the target distribution or vice-versa."""
+    if self.is_bidirectional:
+      return super().transport(vec, forward)
+
+    vec = jnp.atleast_2d(vec)
+    assert (forward
+            is True), ("Only forward mapping (source -> target) is supported.")
+    return self._grad_f(vec)
+
+  def tree_flatten(self) -> Tuple[Sequence[Any], Dict[str, Any]]:
+    """Flatten the kwargs."""
+    return [], {
+        "grad_f": self._grad_f,
+        "g": self._g,
+        "cost_fn": self.cost_fn,
+        "is_bidirectional": self.is_bidirectional,
+        "corr": self._corr
+    }
+
+
+class PotentialModelWrapper(BasePotential):
+  """Wrapper class for the neural models.
+
+  Implements a potential value or a vector field.
+
+  Args:
+    model: Network architecture of the potential.
+    add_l2_norm: If `True``, l2 norm is added to the potential.
+    is_potential: Model the potential if ``True``, otherwise
+      model the gradient of the potential.
+  """
+
+  model: nn.Module
+  add_l2_norm: bool
+  is_potential: bool = True
+
+  @nn.compact
+  def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
+    """Apply model and optionally add l2 norm or x."""
+    z: jnp.ndarray = self.model(x)
+
+    if self.is_potential:
+      z = z.squeeze()
+
+    if self.add_l2_norm:
+      z = z + (0.5 * jnp.dot(x, x)) if self.is_potential else x
+
+    return z
+
+  def potential_gradient_fn(
+      self, params: FrozenDict[str, jnp.ndarray]
+  ) -> PotentialGradientFn_t:
+    """A vector function or gradient of the potential."""
+    if self.is_potential:
+      return jax.grad(self.potential_value_fn(params))
+    return lambda x: self.apply({"params": params}, x)
+
+
+class MLP(nn.Module):
```

Comment:
I'd either:
1. move this to `ott.neural.networks.potentials`
2. or remove this + the default networks in `ExpectileNeuralDual` and just add it to the notebook as a reference potential.

Reply:
Moved to `ott.neural.networks.potentials` to be consistent with `W2NeuralDual` .